src/inmanta/server/services/environment_metrics_service.py

```diff
@@ -0,0 +1,144 @@
+    @abc.abstractmethod
+    def get_metric_name(self) -> str:
+        """
+        Returns the name of the metric collected by this MetricsCollector.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_metric_type(self) -> MetricType:
+        """
+        Returns the type of Metric collected by this metrics collector (count, non-count, etc.).
+        This information is required by the `EnvironmentMetricsService` to know how the data
+        should be aggregated.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    async def get_metric_value(self, start_interval: datetime, end_interval: datetime) -> dict[str, int]:
+        """
+        Invoked by the `EnvironmentMetricsService` at the end of the metrics collection interval.
+        Returns the metrics collected by this MetricCollector within the past metrics collection interval.
+
+        The *_interval arguments are present, because this method is intended to perform a query on
+        the database. No in-memory state is being stored by this metrics collector. The provided interval
+        should be interpreted as [start_interval, end_interval[
+
+        :param start_interval: The start time of the metrics collection interval (inclusive).
+        :param end_interval: The end time of the metrics collection interval (exclusive).
+        :result: The metrics collected by this MetricCollector within the past metrics collection interval.
+        """
+        raise NotImplementedError()
+
+
+class EnvironmentMetricsService(protocol.ServerSlice):
+    """Slice for the management of metrics"""
+
+    def __init__(self) -> None:
+        super(EnvironmentMetricsService, self).__init__(SLICE_ENVIRONMENT_METRICS)
+        self.metrics_collectors: Dict[str, MetricsCollector] = {}
+        self.previous_timestamp = datetime.now()
+
+    def get_dependencies(self) -> List[str]:
+        return [SLICE_DATABASE]
+
+    def get_depended_by(self) -> List[str]:
+        return [SLICE_TRANSPORT]
+
+    async def start(self) -> None:
+        await super().start()
+        self.schedule(self.flush_metrics, COLLECTION_INTERVAL_IN_SEC, initial_delay=0, cancel_on_stop=True)
+
+    def register_metric_collector(self, metrics_collector: MetricsCollector) -> None:
+        """
+        Register the given metrics_collector.
+        """
+        if metrics_collector.get_metric_name() not in self.metrics_collectors:
+            self.metrics_collectors[metrics_collector.get_metric_name()] = metrics_collector
+        else:
+            raise Exception(f"There already is a metric collector with the name {metrics_collector.get_metric_name()}")
+
+    async def flush_metrics(self) -> None:
+        """
+        Invoked at the end of the metrics collection interval. Writes the metrics
+        collected by the MetricsCollectors in the past metrics collection interval
+        to the database.
+        """
+        now: datetime = datetime.now()
```

Comment:
it might be required to use a single database connection for everything here, to prevent database pool starvation. 

Currently, this implies passing the connection as an argument everywhere.